worker/rhpv2.go

```diff
@@ -583,30 +583,39 @@ func (s *Session) Delete(ctx context.Context, sectorIndices []uint64, price type
 //
 // Note that it is typically not necessary to explicitly unlock a contract; the
 // host will do so automatically when the connection closes.
-func (s *Session) Unlock() (err error) {
+func (s *Session) Unlock(ctx context.Context) (err error) {
 	defer wrapErr(&err, "Unlock")
 	s.revision = rhpv2.ContractRevision{}
 	s.key = nil
-	return s.transport.WriteRequest(rhpv2.RPCUnlockID, nil)
+
+	return s.withTransport(ctx, func(transport *rhpv2.Transport) error {
+		return transport.WriteRequest(rhpv2.RPCUnlockID, nil)
+	})
 }
 
 // Close gracefully terminates the session and closes the underlying connection.
-func (s *Session) Close() (err error) {
+func (s *Session) Close(ctx context.Context) (err error) {
```

Comment:
This is pretty nasty. I see why you did it but it really makes me consider replacing the `withTransaction` field with a `withTransaction` method on the session and calling `s.transaction.Close` again.